Panic on going to definition from helix

### Check if applicable

- [x] I have searched the existing issues (**required**)
- [ ] I'm willing to help fix the problem and contribute a pull request

### Describe the bug

I can create link but `gd` (go to definition) panics:
```
2026-05-07T09:28:48.810 helix_lsp::transport [ERROR] zk-lsp err <- "panic: runtime error: invalid memory address or nil pointer dereference\n"
2026-05-07T09:28:48.810 helix_lsp::transport [ERROR] zk-lsp err <- "[signal SIGSEGV: segmentation violation code=0x1 addr=0x8 pc=0x83cf99]\n"
2026-05-07T09:28:48.810 helix_lsp::transport [ERROR] zk-lsp err <- "\n"
2026-05-07T09:28:48.810 helix_lsp::transport [ERROR] zk-lsp err <- "goroutine 50 [running]:\n"
2026-05-07T09:28:48.814 helix_lsp::transport [ERROR] zk-lsp err <- "github.com/zk-org/zk/internal/adapter/lsp.NewServer.func13(0x3348f1804aa0?, 0x3348f1aa8bd0)\n"
2026-05-07T09:28:48.814 helix_lsp::transport [ERROR] zk-lsp err <- "\t/home/alex/code/zk/internal/adapter/lsp/server.go:328 +0x139\n"
2026-05-07T09:28:48.814 helix_lsp::transport [ERROR] zk-lsp err <- "github.com/tliron/glsp/protocol_3_16.(*Handler).Handle(0x3348f197c000, 0x3348f1985040)\n"
2026-05-07T09:28:48.814 helix_lsp::transport [ERROR] zk-lsp err <- "\t/home/alex/go/pkg/mod/github.com/tliron/glsp@v0.1.1/protocol_3_16/handler.go:422 +0x185f\n"
2026-05-07T09:28:48.814 helix_lsp::transport [ERROR] zk-lsp err <- "github.com/tliron/glsp/server.(*Server).handle(0x3348f19280c0, {0xb41448, 0x117f4a0}, 0x3348f17942d0, 0x3348f1aefc20)\n"
2026-05-07T09:28:48.814 helix_lsp::transport [ERROR] zk-lsp err <- "\t/home/alex/go/pkg/mod/github.com/tliron/glsp@v0.1.1/server/handle.go:46 +0x27c\n"
2026-05-07T09:28:48.814 helix_lsp::transport [ERROR] zk-lsp err <- "github.com/sourcegraph/jsonrpc2.(*HandlerWithErrorConfigurer).Handle(0x3348f198a8f0, {0xb41448, 0x117f4a0}, 0x3348f17942d0, 0x3348f1aefc20)\n"
2026-05-07T09:28:48.814 helix_lsp::transport [ERROR] zk-lsp err <- "\t/home/alex/go/pkg/mod/github.com/sourcegraph/jsonrpc2@v0.1.0/handler_with_error.go:21 +0x57\n"
2026-05-07T09:28:48.814 helix_lsp::transport [ERROR] zk-lsp err <- "github.com/sourcegraph/jsonrpc2.(*Conn).readMessages(0x3348f17942d0, {0xb41448, 0x117f4a0})\n"
2026-05-07T09:28:48.814 helix_lsp::transport [ERROR] zk-lsp err <- "\t/home/alex/go/pkg/mod/github.com/sourcegraph/jsonrpc2@v0.1.0/jsonrpc2.go:553 +0x331\n"
2026-05-07T09:28:48.814 helix_lsp::transport [ERROR] zk-lsp err <- "created by github.com/sourcegraph/jsonrpc2.NewConn in goroutine 1\n"
2026-05-07T09:28:48.814 helix_lsp::transport [ERROR] zk-lsp err <- "\t/home/alex/go/pkg/mod/github.com/sourcegraph/jsonrpc2@v0.1.0/jsonrpc2.go:334 +0x1e6\n"
2026-05-07T09:28:48.818 helix_term::commands::lsp [ERROR] Error requesting locations: server closed the stream
```
`zk` is built with
```
CGO_ENABLED=1 go build -buildvcs=false -tags "fts5" -ldflags "-X=main.Version=v0.15.4-15-g0b8e6ae"
```

`$SHELL` is set to `/bin/bash`. Thanks !

### How to reproduce?

1. `zk init` on a empty folder
2. create a link with code action `spc-a`
3. `gd` panics.

`helix` configurations

```
[language-server.zk]
command = "zk"
args = ["lsp"]

[[language]]
name = "markdown"
roots = [".zk"]
language-servers = ["zk"]

```



### zk configuration

```toml
# zk configuration file
#
# Uncomment the properties you want to customize.

# NOTE SETTINGS
#
# Defines the default options used when generating new notes.
[note]
template = "default.md"
[extra]
[format.markdown]

# Enable support for #hashtags.
hashtags = false
# Enable support for :colon:separated:tags:.
colon-tags = false
# Enable support for Bear's #multi-word tags#
# Hashtags must be enabled for multi-word tags to work.
multiword-tags = false


# EXTERNAL TOOLS
[tool]
[lsp]

[lsp.diagnostics]
dead-link = "error"
[lsp.completion]

[filter]

[alias]
```

### Environment

```bash
zk 0.15.4-15-g0b8e6ae
system: Linux 6.18.26-gentoo-dist x86_64 GNU/Linux
```

--- a/internal/adapter/lsp/server.go
+++ b/internal/adapter/lsp/server.go
@@ -325,7 +325,7 @@ func NewServer(opts ServerOpts) *Server {
 			return nil, err
 		}
 
-		if isTrue(clientCapabilities.TextDocument.Definition.LinkSupport) {
+		if definitionLinkSupport(clientCapabilities) {
 			return protocol.LocationLink{
 				OriginSelectionRange: &link.Range,
 				TargetURI:            target.URI,
@@ -934,6 +934,18 @@ func boolPtr(v bool) *bool {
 
 func isTrue(v *bool) bool {
 	return v != nil && *v
+}
+
+// definitionLinkSupport reports whether the client advertised
+// textDocument.definition.linkSupport=true. The intermediate capability
+// objects are pointers and may be nil for clients (e.g. helix) that do
+// not advertise every sub-capability; dereferencing them blindly used
+// to crash the LSP server on textDocument/definition (#716).
+func definitionLinkSupport(caps protocol.ClientCapabilities) bool {
+	if caps.TextDocument == nil || caps.TextDocument.Definition == nil {
+		return false
+	}
+	return isTrue(caps.TextDocument.Definition.LinkSupport)
 }
 
 func stringPtr(v string) *string {

--- a/internal/adapter/lsp/server.go
+++ b/internal/adapter/lsp/server.go
@@ -936,16 +936,9 @@ func isTrue(v *bool) bool {
 	return v != nil && *v
 }
 
-// definitionLinkSupport reports whether the client advertised
-// textDocument.definition.linkSupport=true. The intermediate capability
-// objects are pointers and may be nil for clients (e.g. helix) that do
-// not advertise every sub-capability; dereferencing them blindly used
-// to crash the LSP server on textDocument/definition (#716).
+// definitionLinkSupport reports whether the client advertised textDocument.definition.linkSupport=true.
 func definitionLinkSupport(caps protocol.ClientCapabilities) bool {
-	if caps.TextDocument == nil || caps.TextDocument.Definition == nil {
-		return false
-	}
-	return isTrue(caps.TextDocument.Definition.LinkSupport)
+	return caps.TextDocument != nil && caps.TextDocument.Definition != nil && isTrue(caps.TextDocument.Definition.LinkSupport)
 }
 
 func stringPtr(v string) *string {

--- a/internal/adapter/lsp/server_test.go
+++ b/internal/adapter/lsp/server_test.go
@@ -169,3 +169,50 @@ func TestServer_buildInvokedCompletionList(t *testing.T) {
 		})
 	}
 }
+
+// TestDefinitionLinkSupport_NilCapabilities is a regression test for #716:
+// clients (e.g. helix) that omit `textDocument` or `textDocument.definition`
+// from their initialize-capabilities used to nil-deref the LSP server when
+// handling textDocument/definition. definitionLinkSupport must treat a
+// missing intermediate object as "false" (no LinkSupport) rather than panic.
+func TestDefinitionLinkSupport_NilCapabilities(t *testing.T) {
+	// Whole TextDocument missing.
+	caps := protocol.ClientCapabilities{}
+	assert.Equal(t, definitionLinkSupport(caps), false)
+
+	// TextDocument present but Definition missing (helix's shape).
+	caps = protocol.ClientCapabilities{
+		TextDocument: &protocol.TextDocumentClientCapabilities{},
+	}
+	assert.Equal(t, definitionLinkSupport(caps), false)
+
+	// Definition present but LinkSupport pointer nil.
+	caps = protocol.ClientCapabilities{
+		TextDocument: &protocol.TextDocumentClientCapabilities{
+			Definition: &protocol.DefinitionClientCapabilities{},
+		},
+	}
+	assert.Equal(t, definitionLinkSupport(caps), false)
+
+	// Definition.LinkSupport explicitly false.
+	flag := false
+	caps = protocol.ClientCapabilities{
+		TextDocument: &protocol.TextDocumentClientCapabilities{
+			Definition: &protocol.DefinitionClientCapabilities{
+				LinkSupport: &flag,
+			},
+		},
+	}
+	assert.Equal(t, definitionLinkSupport(caps), false)
+
+	// Definition.LinkSupport explicitly true (the only true path).
+	flag = true
+	caps = protocol.ClientCapabilities{
+		TextDocument: &protocol.TextDocumentClientCapabilities{
+			Definition: &protocol.DefinitionClientCapabilities{
+				LinkSupport: &flag,
+			},
+		},
+	}
+	assert.Equal(t, definitionLinkSupport(caps), true)
+}

--- a/internal/adapter/lsp/server_test.go
+++ b/internal/adapter/lsp/server_test.go
@@ -170,23 +170,15 @@ func TestServer_buildInvokedCompletionList(t *testing.T) {
 	}
 }
 
-// TestDefinitionLinkSupport_NilCapabilities is a regression test for #716:
-// clients (e.g. helix) that omit `textDocument` or `textDocument.definition`
-// from their initialize-capabilities used to nil-deref the LSP server when
-// handling textDocument/definition. definitionLinkSupport must treat a
-// missing intermediate object as "false" (no LinkSupport) rather than panic.
 func TestDefinitionLinkSupport_NilCapabilities(t *testing.T) {
-	// Whole TextDocument missing.
 	caps := protocol.ClientCapabilities{}
 	assert.Equal(t, definitionLinkSupport(caps), false)
 
-	// TextDocument present but Definition missing (helix's shape).
 	caps = protocol.ClientCapabilities{
 		TextDocument: &protocol.TextDocumentClientCapabilities{},
 	}
 	assert.Equal(t, definitionLinkSupport(caps), false)
 
-	// Definition present but LinkSupport pointer nil.
 	caps = protocol.ClientCapabilities{
 		TextDocument: &protocol.TextDocumentClientCapabilities{
 			Definition: &protocol.DefinitionClientCapabilities{},
@@ -194,7 +186,6 @@ func TestDefinitionLinkSupport_NilCapabilities(t *testing.T) {
 	}
 	assert.Equal(t, definitionLinkSupport(caps), false)
 
-	// Definition.LinkSupport explicitly false.
 	flag := false
 	caps = protocol.ClientCapabilities{
 		TextDocument: &protocol.TextDocumentClientCapabilities{
@@ -205,7 +196,6 @@ func TestDefinitionLinkSupport_NilCapabilities(t *testing.T) {
 	}
 	assert.Equal(t, definitionLinkSupport(caps), false)
 
-	// Definition.LinkSupport explicitly true (the only true path).
 	flag = true
 	caps = protocol.ClientCapabilities{
 		TextDocument: &protocol.TextDocumentClientCapabilities{

